File: js/data/coordinate-extract.js
const fs = require('fs');
const https = require('https');
const { createObjectCsvWriter } = require('csv-writer');

// Configuration
const GOOGLE_API_KEY = "AIzaSyAiVn2TbI7qSuTzw1EKvY4urq7V5aTZkZg"; // Replace with your actual API key
const GEOCODING_API_URL = "https://maps.googleapis.com/maps/api/geocode/json";
const ROADS_API_URL = "https://roads.googleapis.com/v1/snapToRoads";

// Chennai region bounds
const CHENNAI_BOUNDS = {
    northeast: { lat: 13.3, lng: 80.5 },
    southwest: { lat: 12.7, lng: 79.8 }
};

class StopCoordinateExtractor {
    constructor(apiKey) {
        this.apiKey = apiKey;
        this.extractedCoordinates = [];
        this.failedStops = [];
    }

    async makeRequest(url) {
        return new Promise((resolve, reject) => {
            https.get(url, (response) => {
                let data = '';
                response.on('data', (chunk) => {
                    data += chunk;
                });
                response.on('end', () => {
                    try {
                        resolve(JSON.parse(data));
                    } catch (error) {
                        reject(error);
                    }
                });
            }).on('error', (error) => {
                reject(error);
            });
        });
    }

    async geocodeStop(stopName, routeName = null) {
        const searchQueries = [
            `${stopName}, Chennai, Tamil Nadu, India`,
            `${stopName}, Chennai`,
            `${stopName} bus stop, Chennai`,
            `${stopName} Chennai Tamil Nadu`
        ];

        if (routeName) {
            searchQueries.unshift(`${stopName}, ${routeName}, Chennai`);
        }

        for (const query of searchQueries) {
            try {
                const encodedQuery = encodeURIComponent(query);
                const bounds = `${CHENNAI_BOUNDS.southwest.lat},${CHENNAI_BOUNDS.southwest.lng}|${CHENNAI_BOUNDS.northeast.lat},${CHENNAI_BOUNDS.northeast.lng}`;
                const url = `${GEOCODING_API_URL}?address=${encodedQuery}&key=${this.apiKey}&region=in&bounds=${bounds}`;

                const data = await this.makeRequest(url);

                if (data.status === 'OK' && data.results.length > 0) {
                    const result = data.results[0];
                    const location = result.geometry.location;

                    if (this.isWithinChennai(location.lat, location.lng)) {
                        console.log(`✅ Found: ${stopName} -> ${location.lat.toFixed(6)}, ${location.lng.toFixed(6)}`);
                        return {
                            stopName: stopName,
                            routeName: routeName,
                            latitude: location.lat,
                            longitude: location.lng,
                            formattedAddress: result.formatted_address,
                            placeId: result.place_id || '',
                            searchQuery: query
                        };
                    }
                }

                // Rate limiting
                await this.sleep(100);

            } catch (error) {
                console.log(`❌ Error geocoding ${stopName} with query '${query}': ${error.message}`);
                continue;
            }
        }

        console.log(`⚠️  Failed to find coordinates for: ${stopName}`);
        return null;
    }

    isWithinChennai(lat, lng) {
        return (CHENNAI_BOUNDS.southwest.lat <= lat <= CHENNAI_BOUNDS.northeast.lat &&
                CHENNAI_BOUNDS.southwest.lng <= lng <= CHENNAI_BOUNDS.northeast.lng);
    }

    async snapToRoad(lat, lng) {
        try {
            const url = `${ROADS_API_URL}?path=${lat},${lng}&key=${this.apiKey}`;
            const data = await this.makeRequest(url);

            if (data.snappedPoints && data.snappedPoints.length > 0) {
                const snapped = data.snappedPoints[0].location;
                console.log(`🛣️  Snapped to road: ${lat.toFixed(6)},${lng.toFixed(6)} -> ${snapped.latitude.toFixed(6)},${snapped.longitude.toFixed(6)}`);
                return [snapped.latitude, snapped.longitude];
            }

        } catch (error) {
            console.log(`⚠️  Road snapping failed for ${lat},${lng}: ${error.message}`);
        }

        return [lat, lng];
    }

    async processRoutesFile(jsonFilePath) {
        console.log(`📂 Loading routes from: ${jsonFilePath}`);

        let data;
        try {
            const fileContent = fs.readFileSync(jsonFilePath, 'utf8');
            data = JSON.parse(fileContent);
        } catch (error) {
            console.log(`❌ Error loading JSON file: ${error.message}`);
            return [];
        }

        const routes = data.routes || [];
        console.log(`📊 Found ${routes.length} routes to process`);

        const allCoordinates = [];
        let totalStops = 0;
        let successfulStops = 0;

        for (let routeIdx = 0; routeIdx < routes.length; routeIdx++) {
            const route = routes[routeIdx];
            const routeName = route.routename || `Route_${routeIdx}`;
            const stops = route.stops || {};

            console.log(`\n🚌 Processing route: ${routeName} (${Object.keys(stops).length} stops)`);

            for (const [stopKey, stopName] of Object.entries(stops)) {
                totalStops++;
                console.log(`   📍 Processing stop ${stopKey}: ${stopName}`);

                const coordData = await this.geocodeStop(stopName, routeName);

                if (coordData) {
                    const [snappedLat, snappedLng] = await this.snapToRoad(
                        coordData.latitude,
                        coordData.longitude
                    );

                    const finalCoordData = {
                        ...coordData,
                        originalLatitude: coordData.latitude,
                        originalLongitude: coordData.longitude,
                        snappedLatitude: snappedLat,
                        snappedLongitude: snappedLng,
                        stopId: stopKey,
                        routeIndex: routeIdx
                    };

                    allCoordinates.push(finalCoordData);
                    successfulStops++;
                } else {
                    this.failedStops.push({
                        routeName: routeName,
                        stopId: stopKey,
                        stopName: stopName
                    });
                }

                await this.sleep(200);
            }
        }

        console.log(`\n📊 Processing Summary:`);
        console.log(`   Total stops processed: ${totalStops}`);
        console.log(`   Successfully geocoded: ${successfulStops}`);
        console.log(`   Failed to geocode: ${this.failedStops.length}`);
        console.log(`   Success rate: ${((successfulStops/totalStops)*100).toFixed(1)}%`);

        return allCoordinates;
    }

    async saveToCSV(coordinates, outputFile) {
        console.log(`\n💾 Saving coordinates to: ${outputFile}`);

        if (!coordinates.length) {
            console.log("❌ No coordinates to save");
            return;
        }

        const csvWriter = createObjectCsvWriter({
            path: outputFile,
            header: [
                { id: 'routeName', title: 'route_name' },
                { id: 'routeIndex', title: 'route_index' },
                { id: 'stopId', title: 'stop_id' },
                { id: 'stopName', title: 'stop_name' },
                { id: 'originalLatitude', title: 'original_latitude' },
                { id: 'originalLongitude', title: 'original_longitude' },
                { id: 'snappedLatitude', title: 'snapped_latitude' },
                { id: 'snappedLongitude', title: 'snapped_longitude' },
                { id: 'formattedAddress', title: 'formatted_address' },
                { id: 'placeId', title: 'place_id' },
                { id: 'searchQuery', title: 'search_query' }
            ]
        });

        try {
            await csvWriter.writeRecords(coordinates);
            console.log(`✅ Successfully saved ${coordinates.length} coordinates to ${outputFile}`);
        } catch (error) {
            console.log(`❌ Error saving to CSV: ${error.message}`);
        }
    }

    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

async function main() {
    console.log("🚀 Starting Stop Coordinate Extraction");
    console.log("=".repeat(50));

    const extractor = new StopCoordinateExtractor(GOOGLE_API_KEY);

    const inputFile = "/Users/mito_1315_/Documents/Workspace/Transport/Routes_Data/Existing-Routes/routes.json";
    const outputFile = "/Users/mito_1315_/Documents/Workspace/Transport/Routes_Data/extracted_stop_coordinates.csv";

    const coordinates = await extractor.processRoutesFile(inputFile);

    if (coordinates.length > 0) {
        await extractor.saveToCSV(coordinates, outputFile);

        console.log(`\n🔍 Sample coordinates:`);
        for (let i = 0; i < Math.min(5, coordinates.length); i++) {
            const coord = coordinates[i];
            console.log(`   ${coord.stopName}: ${coord.snappedLatitude.toFixed(6)}, ${coord.snappedLongitude.toFixed(6)}`);
        }
    }

    console.log(`\n✅ Coordinate extraction completed!`);
    console.log(`📁 Results saved to: ${outputFile}`);
}

if (require.main === module) {
    main().catch(console.error);
}

module.exports = StopCoordinateExtractor;

File: js/data/coordinate-extract.py
import json
import requests
import csv
import time
from typing import Dict, List, Tuple, Optional

# Configuration
GOOGLE_API_KEY = "AIzaSyAiVn2TbI7qSuTzw1EKvY4urq7V5aTZkZg"  # Replace with your actual API key
GEOCODING_API_URL = "https://maps.googleapis.com/maps/api/geocode/json"
ROADS_API_URL = "https://roads.googleapis.com/v1/snapToRoads"

# Chennai region bounds for better geocoding accuracy
CHENNAI_BOUNDS = {
    "northeast": {"lat": 13.3, "lng": 80.5},
    "southwest": {"lat": 12.7, "lng": 79.8}
}

class StopCoordinateExtractor:
    def __init__(self, api_key: str):
        self.api_key = api_key
        self.extracted_coordinates = []
        self.failed_stops = []
        
    def geocode_stop(self, stop_name: str, route_name: str = None) -> Optional[Dict]:
        """
        Geocode a stop name to get coordinates using Google Maps Geocoding API
        """
        # Enhance search query with Chennai context
        search_queries = [
            f"{stop_name}, Chennai, Tamil Nadu, India",
            f"{stop_name}, Chennai",
            f"{stop_name} bus stop, Chennai",
            f"{stop_name} Chennai Tamil Nadu"
        ]
        
        # If route name is provided, try with route context
        if route_name:
            search_queries.insert(0, f"{stop_name}, {route_name}, Chennai")
        
        for query in search_queries:
            try:
                params = {
                    'address': query,
                    'key': self.api_key,
                    'region': 'in',  # India region bias
                    'bounds': f"{CHENNAI_BOUNDS['southwest']['lat']},{CHENNAI_BOUNDS['southwest']['lng']}|{CHENNAI_BOUNDS['northeast']['lat']},{CHENNAI_BOUNDS['northeast']['lng']}"
                }
                
                response = requests.get(GEOCODING_API_URL, params=params)
                data = response.json()
                
                if data['status'] == 'OK' and len(data['results']) > 0:
                    result = data['results'][0]
                    location = result['geometry']['location']
                    
                    # Check if the result is within Chennai bounds
                    if self.is_within_chennai(location['lat'], location['lng']):
                        print(f"✅ Found: {stop_name} -> {location['lat']:.6f}, {location['lng']:.6f}")
                        return {
                            'stop_name': stop_name,
                            'route_name': route_name,
                            'latitude': location['lat'],
                            'longitude': location['lng'],
                            'formatted_address': result['formatted_address'],
                            'place_id': result.get('place_id', ''),
                            'search_query': query
                        }
                
                # Rate limiting
                time.sleep(0.1)
                
            except Exception as e:
                print(f"❌ Error geocoding {stop_name} with query '{query}': {str(e)}")
                continue
        
        print(f"⚠️  Failed to find coordinates for: {stop_name}")
        return None
    
    def is_within_chennai(self, lat: float, lng: float) -> bool:
        """Check if coordinates are within Chennai bounds"""
        return (CHENNAI_BOUNDS['southwest']['lat'] <= lat <= CHENNAI_BOUNDS['northeast']['lat'] and
                CHENNAI_BOUNDS['southwest']['lng'] <= lng <= CHENNAI_BOUNDS['northeast']['lng'])
    
    def snap_to_road(self, lat: float, lng: float) -> Tuple[float, float]:
        """
        Snap coordinates to nearest road using Google Roads API
        """
        try:
            params = {
                'path': f"{lat},{lng}",
                'key': self.api_key
            }
            
            response = requests.get(ROADS_API_URL, params=params)
            data = response.json()
            
            if 'snappedPoints' in data and len(data['snappedPoints']) > 0:
                snapped = data['snappedPoints'][0]['location']
                print(f"🛣️  Snapped to road: {lat:.6f},{lng:.6f} -> {snapped['latitude']:.6f},{snapped['longitude']:.6f}")
                return snapped['latitude'], snapped['longitude']
            
        except Exception as e:
            print(f"⚠️  Road snapping failed for {lat},{lng}: {str(e)}")
        
        # Return original coordinates if snapping fails
        return lat, lng
    
    def process_routes_file(self, json_file_path: str) -> List[Dict]:
        """
        Process the routes.json file and extract coordinates for all stops
        """
        print(f"📂 Loading routes from: {json_file_path}")
        
        try:
            with open(json_file_path, 'r', encoding='utf-8') as file:
                data = json.load(file)
        except Exception as e:
            print(f"❌ Error loading JSON file: {str(e)}")
            return []
        
        routes = data.get('routes', [])
        print(f"📊 Found {len(routes)} routes to process")
        
        all_coordinates = []
        total_stops = 0
        successful_stops = 0
        
        for route_idx, route in enumerate(routes):
            route_name = route.get('routename', f'Route_{route_idx}')
            stops = route.get('stops', {})
            
            print(f"\n🚌 Processing route: {route_name} ({len(stops)} stops)")
            
            for stop_key, stop_name in stops.items():
                total_stops += 1
                print(f"   📍 Processing stop {stop_key}: {stop_name}")
                
                # Geocode the stop
                coord_data = self.geocode_stop(stop_name, route_name)
                
                if coord_data:
                    # Snap to road for better accuracy
                    snapped_lat, snapped_lng = self.snap_to_road(
                        coord_data['latitude'], 
                        coord_data['longitude']
                    )
                    
                    # Update with snapped coordinates
                    coord_data.update({
                        'original_latitude': coord_data['latitude'],
                        'original_longitude': coord_data['longitude'],
                        'snapped_latitude': snapped_lat,
                        'snapped_longitude': snapped_lng,
                        'stop_id': stop_key,
                        'route_index': route_idx
                    })
                    
                    all_coordinates.append(coord_data)
                    successful_stops += 1
                else:
                    self.failed_stops.append({
                        'route_name': route_name,
                        'stop_id': stop_key,
                        'stop_name': stop_name
                    })
                
                # Rate limiting to avoid API quota issues
                time.sleep(0.2)
        
        print(f"\n📊 Processing Summary:")
        print(f"   Total stops processed: {total_stops}")
        print(f"   Successfully geocoded: {successful_stops}")
        print(f"   Failed to geocode: {len(self.failed_stops)}")
        print(f"   Success rate: {(successful_stops/total_stops)*100:.1f}%")
        
        return all_coordinates
    
    def save_to_csv(self, coordinates: List[Dict], output_file: str):
        """
        Save extracted coordinates to CSV file
        """
        print(f"\n💾 Saving coordinates to: {output_file}")
        
        if not coordinates:
            print("❌ No coordinates to save")
            return
        
        fieldnames = [
            'route_name', 'route_index', 'stop_id', 'stop_name',
            'original_latitude', 'original_longitude',
            'snapped_latitude', 'snapped_longitude',
            'formatted_address', 'place_id', 'search_query'
        ]
        
        try:
            with open(output_file, 'w', newline='', encoding='utf-8') as csvfile:
                writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
                writer.writeheader()
                
                for coord in coordinates:
                    writer.writerow(coord)
            
            print(f"✅ Successfully saved {len(coordinates)} coordinates to {output_file}")
            
        except Exception as e:
            print(f"❌ Error saving to CSV: {str(e)}")
    
    def save_failed_stops(self, output_file: str):
        """
        Save failed stops to CSV for manual review
        """
        if not self.failed_stops:
            print("✅ No failed stops to save")
            return
        
        print(f"💾 Saving failed stops to: {output_file}")
        
        try:
            with open(output_file, 'w', newline='', encoding='utf-8') as csvfile:
                writer = csv.DictWriter(csvfile, fieldnames=['route_name', 'stop_id', 'stop_name'])
                writer.writeheader()
                
                for failed in self.failed_stops:
                    writer.writerow(failed)
            
            print(f"⚠️  Saved {len(self.failed_stops)} failed stops to {output_file}")
            
        except Exception as e:
            print(f"❌ Error saving failed stops: {str(e)}")
    
    def create_summary_report(self, coordinates: List[Dict]) -> Dict:
        """
        Create a summary report of the extraction process
        """
        if not coordinates:
            return {}
        
        # Group by route
        routes_summary = {}
        for coord in coordinates:
            route_name = coord['route_name']
            if route_name not in routes_summary:
                routes_summary[route_name] = {
                    'total_stops': 0,
                    'stops': [],
                    'avg_lat': 0,
                    'avg_lng': 0
                }
            
            routes_summary[route_name]['total_stops'] += 1
            routes_summary[route_name]['stops'].append(coord['stop_name'])
        
        # Calculate average coordinates for each route
        for route_name in routes_summary:
            route_coords = [c for c in coordinates if c['route_name'] == route_name]
            if route_coords:
                avg_lat = sum(c['snapped_latitude'] for c in route_coords) / len(route_coords)
                avg_lng = sum(c['snapped_longitude'] for c in route_coords) / len(route_coords)
                routes_summary[route_name]['avg_lat'] = avg_lat
                routes_summary[route_name]['avg_lng'] = avg_lng
        
        return {
            'total_routes': len(routes_summary),
            'total_stops_extracted': len(coordinates),
            'total_failed': len(self.failed_stops),
            'routes_summary': routes_summary
        }

def main():
    print("🚀 Starting Stop Coordinate Extraction")
    print("=" * 50)
    
    # Initialize extractor
    extractor = StopCoordinateExtractor(GOOGLE_API_KEY)
    
    # File paths
    input_file = "/Users/mito_1315_/Documents/Workspace/Transport/Routes_Data/Existing-Routes/routes.json"
    output_file = "/Users/mito_1315_/Documents/Workspace/Transport/Routes_Data/extracted_stop_coordinates.csv"
    failed_file = "/Users/mito_1315_/Documents/Workspace/Transport/Routes_Data/failed_stops.csv"
    
    # Process the routes file
    coordinates = extractor.process_routes_file(input_file)
    
    # Save results
    if coordinates:
        extractor.save_to_csv(coordinates, output_file)
        
        # Create summary report
        summary = extractor.create_summary_report(coordinates)
        print(f"\n📊 Final Summary:")
        print(f"   Total routes processed: {summary.get('total_routes', 0)}")
        print(f"   Total stops extracted: {summary.get('total_stops_extracted', 0)}")
        print(f"   Total failed: {summary.get('total_failed', 0)}")
        
        # Save sample of coordinates for verification
        print(f"\n🔍 Sample coordinates:")
        for i, coord in enumerate(coordinates[:5]):
            print(f"   {coord['stop_name']}: {coord['snapped_latitude']:.6f}, {coord['snapped_longitude']:.6f}")
    
    # Save failed stops for manual review
    extractor.save_failed_stops(failed_file)
    
    print(f"\n✅ Coordinate extraction completed!")
    print(f"📁 Results saved to: {output_file}")
    if extractor.failed_stops:
        print(f"📁 Failed stops saved to: {failed_file}")

if __name__ == "__main__":
    main()
